src/main/java/net/minecraftforge/client/model/ModelDataManager.java

```diff
@@ -0,0 +1,101 @@
+package net.minecraftforge.client.model;
+
+import java.lang.ref.WeakReference;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+import com.google.common.base.Preconditions;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+
+import net.minecraft.client.Minecraft;
+import net.minecraft.tileentity.TileEntity;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.ChunkPos;
+import net.minecraft.world.World;
+import net.minecraftforge.client.model.data.IModelData;
+
+public class ModelDataManager
+{
+    private static WeakReference<World> currentWorld = new WeakReference<>(null);
+    
+    private static final Map<ChunkPos, Set<BlockPos>> needModelDataRefresh = new HashMap<>();
```

Comment:
why not just synchronizedMap or concurrent map?
also this name seems a bit weird to me, maybe `dirtyPositions` or `dirtyModelData`?

Reply:
It doesn't need to be threadsafe, it should only be accessed on the client thread. Renames will be considered, it is a bit wordy, but in the end it's an internal field.

Reply:
why is it synchronized on later in this class if it doesn't need to be threadsafe?

Reply:
Just my two cents - the "need" prefix is a question word, implying a boolean (e.g. a flag). It may be just an internal field, but one should care about their fellow programmer (those in future who need to understand and maintain it) :)

Reply:
my question still stands, because as it is the synchronization done is inadequate. if a field is synchronized on, *every* single use of it needs to be synchronized if it has a possibility of running in parallel.

Reply:
You are correct, the call to `cleanCaches` was in a bad spot. None of the current calls to these methods are off-thread, but I do need to support that case.